### DISCUSSION TITLE
Splenic volume approximation using Deep Learning

### DISCUSSION THREAD

**Discussion by Meddebma** [Project-MONAI/MONAI#1698] [Ideas]
Hi, 
I used the spleen segmentation algorithm to segment the decathlon and my own dataset, and wanted to know if I could integrate a function for measuring the splenic volume (in mm3 or mL).
Thanks!

**Comment by Nic-Ma** [comment]
Hi @rijobro and @wyli ,

Could you please help share some comments?

Thanks.

**Comment by rijobro** [comment]
Hi, if you're only doing binary segmentation then you can simply count the number of voxels greater than 0. You could add a new function to your post-transforms, something along the lines of (completely untested):

```python
class GetVol(MapTransform):
    def __init__(keys, select_fn = lambda x: (x>0).sum(), output_key_suffix="_vol"):
        super().__init__(keys)
        self.select_fn = select_fn
        self.output_key_suffix = output_key_suffix
    def __call__(data):
        d = dict(data)
        for key in self.key_iterator(d):
            vol = self.select_fn(d[key])
            d[key + self.output_key_suffix] = vol
        return d
```

If your data is a simple tensor with batch (i.e., `[B,C,H,W,[D]]`, then you could just do (again, untested):

```python
vols = [(b>0).sum() for b in inferred_seg_batch]
```

Once you have the number of voxels, you simply multiply on the spacing per voxel to get the volume.

**Reply by Meddebma** [reply]
Hi @rijobro, 
thanks for helping! 
I used the formula for the batch resulting from inference and I've got this for one segmentation:

Shape: torch.Size([1, 1, 250, 199, 231])
Voxels: [tensor(11491555, device='cuda:0')]

I'm not sure to understand how I can get the spacing per voxel as you meant to calculate to volume
Thanks again!

**Reply by rijobro** [reply]
I wonder if there might be an error in your calculation, since 11491555 is 99.99% of 11492250 (250 x 199 x 231). Anyway, once you've correctly calculated the number of voxels in your segmentation, you multiply by the spacing per voxel. If, for example, you have nifti images and you load them with `LoadImaged`, then their metadata including spacing will be stored. e.g.:

```python
fname_dict = {"image": "/some/path/to/image.nii.gz"}
data = LoadImaged("image")(fname_dict)
spacing = data["image_meta_dict"]["pixdim"][1:]
```

for a 3d image, volume per voxel would be:
```python
spacing = data["image_meta_dict"]["pixdim"][1:4].prod()
```

**Reply by Meddebma** [reply]
Hi @rijobro 
you are right, the calculation is wrong, here is what I did: 

```python
with torch.no_grad():
        metric_sum = 0.0
        metric_count = 0
        for val_data in val_loader:
            # define sliding window size and batch size for windows inference
            val_images, val_labels = val_data["image"].to(device), val_data["label"].to(device)
            roi_size = (96, 96, 96)
            sw_batch_size = 4
            val_outputs = sliding_window_inference(val_data["image"].to(device), roi_size, sw_batch_size, model)
            val_outputs = post_trans(val_outputs)
            vols = [(b>0).sum() for b in val_outputs]
            print(vols)
```
does it make sense?

and for the spacing, I used this transform:
```python
val_transforms = Compose(
            [
                LoadImaged(keys=["image", "label"]),
                AddChanneld(keys=["image", "label"]),
                Spacingd(
                    keys=["image", "label"],
                    pixdim=(1.5, 1.5, 2.0),
                    mode=("bilinear", "nearest"),
                ),
                Orientationd(keys=["image", "label"], axcodes="RAS"),
                ScaleIntensityRanged(
                    keys=["image"], a_min=-57, a_max=164,
                    b_min=0.0, b_max=1.0, clip=True,
                ),
                CropForegroundd(keys=["image", "label"], source_key="image"),
                ToTensord(keys=["image", "label"]),
            ]
        ) 
```
I am wondering if I should take the original voxel spacing or use the pixdim of Spacingd(1.5, 1.5, 2.0)

Thanks

**Reply by rijobro** [reply]
`(1.5, 1.5, 2.0)` is the correct spacing to use. For calculating the number of voxels in the segmentation, that'll depend on `post_trans`. `val_output` needs to be a binary mask, so you'll need to check that's the case.

**Reply by Meddebma** [reply]
Yes in fact it is binary I tried the formula vols = [(b>1).sum() for b in val_outputs] and I've got 0, so every point should be a 0 or 1. That is why I dont get it why its making the voxels of the image not of the segmentation mask

here is my code, it is the last part evaluation with other dataset:
https://github.com/Meddebma/pyradiomics/blob/master/Spleen0221.ipynb

**Reply by Meddebma** [reply]
Hi @rijobro, 

should I mabye add AsDiscrete Transform to postprocessing? would that helps?

Thank you so much

**Reply by rijobro** [reply]
I just tried running your notebook in google colab but it hit some errors. if you'd like me to run it, could you make sure that it works in colab?

**Reply by Meddebma** [reply]
Hi @rijobro,

I changed the formula to get the pixels of the mask:
vols = [(b[:,1]==1).sum() for b in val_outputs]
it worked :) 

thank you very much!
